### ISSUE TITLE
About the camera coordinate type

### ISSUE THREAD

**Issue by Haru2022** [tyz1030/neuralsea#4]
Hi, many thanks for your contribution!
I want to ask that is the camera coordinate type the opengl or opencv type?  The trained images of my own dataset are all totally black, so I doubt that there may be some problems about the coordiante transformation from the world coordiante to the pixel cooridnate.

**Comment by tyz1030** [comment]
working on this. planning to push an update today that, hopefully, can help

**Comment by Haru2022** [comment]
Great! by the way, I'm wondering if the point samping strategy does not consider in the near and far distance info? cuz I check the raysampler.py where the min_depth and the max_depth params are both not used.

**Comment by tyz1030** [comment]
Hi Haru2022,

Thank you for your interests and feedbacks! 

short answer is opencv. xyz is right, down and forward. So when I sample points, z, or say depth, is always positive.

Longer answer if you are using real images and poses from colmap:
I just pushed an update which have one real world data I collected in my watertank. The poses are generated from colmap, following the exact procedure in NeRF (https://github.com/bmild/nerf) and LLFF (https://github.com/Fyusion/LLFF). My code just take the output from the LLFF's img2poses.py script, and somehow I figured out how to bridge it into pytorch3D's convention (check nerf/dataset.py in this repo to see how it works).

min_depth and max_depth were historical junks. I removed them in this update.

There are still several magic numbers that can get things wrong, so I have to point it out to you:
for example, in real_watertank.yaml,   
* image_down_scale: 20: this is because I down scaled the raw camera image from 6000*4000 to 300*200. so I will have to down scale the image size and focal length etc in the code. (not for synthetic data)
* pose_down_scale: 16: This is because the colmap output does not have scale (this is universal monocular vision problem) So I manually scaled the colmap poses down to fit in my iNGP grids. If you don't need any down scaling, set these values to 1. (again, I don't have this for synthetic data as I know the gt poses)
*  near_bounding_in_z: -0.2: This is where near bounding plane starts in global coordinate
*  near_to_far_range_in_z: 0.4: This is the range of near and far bounding plane covers, in z direction, in global coordinate

Not sure if any of these will help you solve you problem. See if you can run my example real/synthetic examples first and let me know. Good luck with your project!

**Comment by Haru2022** [comment]
Hi tyz1030,

I'm still puzzled about the ray construction process in your updated code.

In the raysampler.py, the ray is sampled as below:

origin = -torch.matmul(cameras.T, torch.transpose(cameras.R[0], 0, 1))
        z_to_center = origin[0, -1]
        depths = torch.linspace(
            0, 1, self.n_pts_per_ray, device=cameras.device)
        z_to_foreground = -z_to_center+self.near_bounding_in_z
        depths = z_to_foreground + depths*self.near_to_far_range_in_z

In my understanding, the origin is the t_c_w which is the camera 3d position in the world coordinate. Then why the z value of the cam in the world coords is added into the param z_to_foreground? Becuase in the original nerf, the depth is defined as the depth along the z direction in the camera coordinates while not the world coordinates, which is described as below:

ray = origin_w + normalized_direction_w * depth_in_the_cam_coords

**Comment by Haru2022** [comment]
Hi tyz1030,

I read your updated code especially the raysampler.py. However the ray construction process still confused me a little.

the corresponding code is as below:

origin = -torch.matmul(cameras.T, torch.transpose(cameras.R[0], 0, 1)) 
        z_to_center = origin[0, -1]
        depths = torch.linspace(
            0, 1, self.n_pts_per_ray, device=cameras.device)
        z_to_foreground = -z_to_center+self.near_bounding_in_z
        depths = z_to_foreground + depths*self.near_to_far_range_in_z

In my understanding, <origin = -torch.matmul(cameras.T, torch.transpose(cameras.R[0], 0, 1)) > is the 3d pos of the cam in the world coordinates, which is also called t_c2w. Could you please give me more details about why you add the z value of the camera 3d position in the world coordiantes into the param  z_to_foreground? Because in NeRF, the corresponding construction process is like this:

pts = origins_w + dir_w * depth
depth = self.near_bounding_in_z + length_ratio * (self.near_to_far_range_in_z)

The z value in the world coords seems has no relationship with this process.

**Comment by Haru2022** [comment]
I thought that the reason is that the marching path of the camera of the dataset you provide is just nearly parallel to the z direction, am I right?

**Comment by tyz1030** [comment]
That's right. In early NeRF study people put data into two classes: forward looking and 360 degree. The algorithms are correspondingly different, e.g. NDC is only used in forward looking data ...

While the idea propsed in this work is not limited to certain type of data, our code is only optimized for the forward looking image sequences. My camera on the underwater robot only look into one direction, and my water tank in lab is also too small for me to take photos from more diversified views.

The difference between the sampling methon you mentioned, and my implementation, is that your mentioned method sets bounding plane in each cameras own coordinate system, while mine sets the same bounding plane in the global coordinate. If all cameras are exactly parallel, and you trust the near/far bounding plane estimated from COLMAP, then I believe your method is better for sampling. If there some noises in COLMAP estimated bounding plane (This is rare in air but more likely to happen in scatterring medium like turbid water), or cameras are not very parallel between each other, your viewing frustum won't overlap very well. Which means some poses are well contrained by multiple-view geometry, while some sample points from other poses are only observed in single/few views (visdom visualize this). In such cases, nn will find it easier to overfit given those underconstrained sample points. That's why I decided to put the bounding plane in the global coordinate, so make sure at least the center part of the scene is well constrained. The drawback is that I need to manually set up the near bounding plane in z, and that global z axis need to align, roughly, with all camera's z axis.

Hope this answer make some sense to you, and if any question do not hesitate to let me know.

**Comment by Haru2022** [comment]
1.Thanks for your kindly explaination. I have no further question about it. I test the watertank dataset and the training result is consistent with what you show in the readme.md. You've done a really nice work!

2.But I just find that the coordinate transformation process of the sampled rays may not be correct. The original code is as below.
```python
            #!/usr/bin/env python
            xy_coords = torch.tensor(
                    [[self.image_width, self.image_height]])*temp_xys/2/cameras.focal_length.cpu()

                ray_dirs_cam = torch.cat(
                    [xy_coords, torch.ones_like(temp_xys)[:, 0, None]], dim=1)
                ray_dirs_world = torch.matmul(
                    ray_dirs_cam.cuda(), cameras.R[0].transpose(0, 1))
                rays_dirs_world_normalized = ray_dirs_world / \
                    torch.abs(ray_dirs_world[:, 2])[..., None]
                rays_zs = depths[None, None].expand(
                    batch_size, self.n_rays_per_image, self.n_pts_per_ray)
```

 However, there's no need, in my opinion, to normalize the ray_dirs_world because the ray_dirs_cam already represents the length ratio of $x,y,z$ value of the sampled points along the rays, and the z value is already set to 1 which means the unit depth. The only thing we need to do is to just scale the direction vector with unit depth and transform it to the world coordinate. Because the depth is a scalar, we can transform the rays_dirs_cam to the world coordinate first and then multiply ray_dirs_world with the depth. So I think the rays_dirs_world_normalized may be unneccessary.

I find this possible problem when I use my own dataset and modify the code to set bounding plane in each camera's own coordinate system we discuss before. The max depth of the sampling points is 25m, and the x,y coordinate range of the cams is around within [-50,60], [-60,60]. However, the coordinate of the sampled points along rays always beyond [-200,200] which is obviously unormal.

3.Just another small problem: when the depths list contains zero (e.g., when z_to_foreground is 0), the training loss will become nan (:

**Comment by tyz1030** [comment]
1. Thanks so much for your recognition. Your comments and suggestions are truelly invaluable to us.

2. Agreed! If you want to place bounding planes in camera's coordinate, then normalization by z value (what I did in my code) will be wrong. But, you may normalize it by l2 norm so it has unit length (optional).
But if I'm sticking to my sampling strategy (bounding by z in global coordinate), then my code you mentioned above is actually necessary I believe, since the z axis after torch.matmul() operation won't be the same z axis. I need to normalize by the new z axis.

3. Thanks for pointing out. A lot of such numerical errors I've encountered. A little bit tired of them.. set it to 0.01, then ignore it. :D

overall thank you for the feedbacks if anything unclear let me know at any time

**Comment by Haru2022** [comment]
Cons to you about the acceptance of this paper!!!

**Comment by tyz1030** [comment]
Thank you !
